### PR TITLE
Add hallucination multicalibrator, with example benchmark

### PR DESCRIPTION
**Pull request by gianlucadetommaso**
Add the `HallucinationMulticalibrator`. This is able to take in input a generative model, a tokenizer and some data, then fit the calibrator and predict calibrated probabilities of hallucinations.

### CODE REVIEW

File: benchmarks/hallucination/mmlu/run.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    )
+
+    status = calibrator.fit(
+        texts=calib_choices,
+        contexts=calib_questions,
+        targets=calib_targets,
+    )
+
+    with open("fitted_calibrator.pth", "wb") as filehandler:
+        pickle.dump(calibrator, filehandler, -1)
```
Comment (wistuba) [inline benchmarks/hallucination/mmlu/run.py:84]: is this important? do we need to load that later?

File: fortuna/hallucination/base.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        dataset.
+        Under the hood, the method constructs model-based low-dimensional embeddings of the initial contexts plus texts,
+        it clusters them to provide meaningful (possibly overlapping) groups,
+        it computes the scores for each text,
+        and it finally calibrates the scores on each subset formed by the intersection of each group and each level set
+        of the score function.
+
+        Parameters
+        ----------
+        generative_model: nn.Module
```
Comment (wistuba) [inline fortuna/hallucination/base.py:47]: there is tooling to automatically infer typing from the type hints for the documentation. might save you some time and ensure consistency

File: fortuna/hallucination/base.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+from fortuna.hallucination.grouping.clustering.base import GroupingModel
+from fortuna.hallucination.scoring.inv_perplexity import inv_perplexity
+
+
+class HallucinationMulticalibrator:
+    def __init__(
+        self,
+        generative_model: nn.Module,
+        tokenizer: PreTrainedTokenizer,
+        embedding_reduction_fn: Callable[[np.ndarray], np.ndarray] = None,
```
Comment (wistuba) [inline fortuna/hallucination/base.py:28]: Optional

File: fortuna/hallucination/base.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+            A list of clustering models.
+        scoring_fn: Callable[[torch.Tensor, torch.Tensor, int], torch.Tensor]
+            A scoring function.
+        """
+        self.generative_model = generative_model
+        self.tokenizer = tokenizer
+        if embedding_reduction_fn is not None:
+            self.embedding_reduction_fn = embedding_reduction_fn
+        else:
+            self.embedding_reduction_fn = locally_linear_embedding_fn
```
Comment (wistuba) [inline fortuna/hallucination/base.py:63]: `self._embedding_reduction_fn = embedding_reduction_fn or locally_linear_embedding_fn` if you like it

File: fortuna/hallucination/base.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+
+class HallucinationMulticalibrator:
+    def __init__(
+        self,
+        generative_model: nn.Module,
+        tokenizer: PreTrainedTokenizer,
+        embedding_reduction_fn: Callable[[np.ndarray], np.ndarray] = None,
+        clustering_models: Optional[List] = None,
+        scoring_fn: Callable[
```
Comment (wistuba) [inline fortuna/hallucination/base.py:30]: embedding_reduction_fn and scoring_fn deal differently with default settings. why?

File: fortuna/hallucination/base.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+                    _text_inputs = self.tokenizer(_text, return_tensors="pt").to(
+                        self.generative_model.device
+                    )
+                    _inputs = {
+                        k: torch.cat((context_inputs[k], v), dim=1)
+                        for k, v in _text_inputs.items()
+                    }
+
+                    with torch.no_grad():
+                        __logits = self.generative_model(
```
Comment (wistuba) [inline fortuna/hallucination/base.py:273]: double `_` intentional?

File: fortuna/hallucination/base.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        scores = []
+        embeddings = []
+        which_choices = []
+
+        for text, context in tqdm(zip(texts, contexts)):
+            context_inputs = self.tokenizer(context, return_tensors="pt").to(
+                self.generative_model.device
+            )
+            len_context_inputs = len(context_inputs)
+            if isinstance(text, list):
```
Comment (wistuba) [inline fortuna/hallucination/base.py:259]: correct me if I am wrong, but you have here 2 cases: 1) text is a list of strings 2) text is a string. In the former case, you run over the list and apply some logic L. In the latter case, you also run the same logic but on the text directly. Can you remove the redundant code by something like
```
if isinstance(text, str):
    text = [text]
```
or maybe even better, move the logic to a function?

File: fortuna/hallucination/base.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+                        logits=_logits,
+                        labels=inputs["input_ids"],
+                        init_pos=len_context_inputs,
+                    )
+                    .cpu()
+                    .numpy()
+                )
+
+            else:
+                raise ValueError(
```
Comment (wistuba) [inline fortuna/hallucination/base.py:321]: after refactoring, you might ignore this case and let the user run into the exception. you can't protect against all wrong inputs in this way

File: tests/fortuna/hallucination/scoring.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+from jax import random
+import numpy as np
+import torch
+
+from fortuna.data import InputsLoader
+from fortuna.hallucination.scoring.inv_perplexity import inv_perplexity
+
+
+class TestScoringModel(unittest.TestCase):
+    def test_score(self):
```
Comment (wistuba) [inline tests/fortuna/hallucination/scoring.py:12]: can't you compare against exact values?

File: fortuna/hallucination/utils/string.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import re
+
+
+def string_cleaner(text: str) -> str:
```
Comment (wistuba) [inline fortuna/hallucination/utils/string.py:4]: great candidate for a function to have unit tests

**Review by wistuba — COMMENTED** [review]
I left some comments. Maybe you want to add some more unit tests.

File: benchmarks/hallucination/mmlu/run.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            load_dataset(
+                "lukaemon/mmlu",
+                task,
+            ),
+            task,
+        )
+        for task in task_list
+    ]
+
+    answer_map = {a: i for i, a in enumerate(auc)}
```
[the remark this answers is not in the document]
Comment (gianlucadetommaso) [reply benchmarks/hallucination/mmlu/run.py:40]: Tries to be more generic, but you're right, can be restricted to "ABCD".

File: benchmarks/hallucination/mmlu/run.py (hunk 1 above)
Comment (gianlucadetommaso) [reply benchmarks/hallucination/mmlu/run.py:84]: Yes. We might want to use the calibrated model in a demo where the user brings in their own answers.

File: fortuna/hallucination/base.py (hunk 2 above)
Comment (gianlucadetommaso) [reply fortuna/hallucination/base.py:47]: What tool do you suggest?

File: fortuna/hallucination/base.py (hunk 6 above)
Comment (gianlucadetommaso) [reply fortuna/hallucination/base.py:273]: Yep

File: fortuna/hallucination/base.py (hunk 7 above)
Comment (gianlucadetommaso) [reply fortuna/hallucination/base.py:259]: Yes looks possible, I don't know why I didn't.

**Pull request merged**